### PR TITLE
Build documentation on dev.major without warnings

### LINKED ISSUE #1903: Build documentation without warning on `dev.major`
### Problem Description

The documentation of the branch `dev.major` raise over 50 warnings when building.
These are mostly formatting error in functions and classes doc-strings, but some of the code example also raise errors.
These warnings need to be fixed and the `-W` option to turn warnings into errors, restored.

### Proposed Solution

A recent documentation build can be found [here](https://github.com/qutip/qutip/runs/6596735860?check_suite_focus=true). The problematic doc-string need to be reworded and/or reformatted.

### Alternate Solutions

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Ericgig**
**Description**
Fix documentation so it build without warnings in dev.major and restore the `-W` sphinx options:
- Removed reference to deprecated function and classes.
- Fixed code where functions calls changed.
- Fixed doc-strings format.

However I did not fix documentation to match v5 changes, nor added new functions and classes to the api doc: `qutip.solve.mesolve` is still used for the documentation, not `qutip.solver.mesolve`, etc. 

**Related issues or PRs**
fix #1903

### CODE REVIEW

File: doc/guide/dynamics/dynamics-bloch-redfield.rst [hunk 1]
```diff
@@ -226,15 +226,17 @@ Note that it is also possible to add Lindblad dissipation superoperators in the
 
     H = - delta/2.0 * sigmax() - eps0/2.0 * sigmaz()
 
-    def ohmic_spectrum(w):
+    def ohmic_spectrum(_, w):
```
Comment (hodgestar) [inline doc/guide/dynamics/dynamics-bloch-redfield.rst:229]: Having to specify `_` in order to receive and ignore `t` seems a bit odd, but that is unrelated to this PR.

File: doc/guide/dynamics/dynamics-time.rst [hunk 2]
```diff
@@ -270,38 +270,45 @@ Collapse operators are handled in the exact same way.
...
-    t = np.linspace(-15, 15, 100)
-    func = lambda t: 9*np.exp(-(t / 5)** 2)
-    noisy_func = lambda t: func(t)+(0.05*func(t))*np.random.randn(t.shape[0])
-    noisy_data = noisy_func(t)
+    times = np.linspace(-15, 15, 100)
+    func = lambda t: 9 * np.exp(-(t / 5)** 2)
+    noisy_data = np.array([
+        func(t) * (1 + 0.05 * np.random.randn())
+        for t in times
+    ])
```
Comment (hodgestar) [inline doc/guide/dynamics/dynamics-time.rst:286]: Could we not keep this vectorised as `noisy_data = func(times) * (1 + 0.05 * np.random.randn(len(times)))`?

File: qutip/core/cy/qobjevo.pyx [hunk 3]
```diff
@@ -42,7 +42,7 @@ cdef class QobjEvo:
 
       * a :obj:`~Qobj` (which creates a constant :obj:`~QobjEvo` term).
 
-      * a list of such callables, pairs or :obj:`~Qobj`s.
+      * a list of such callables, pairs or :obj:`~Qobj`\s.
```
Comment (hodgestar) [inline qutip/core/cy/qobjevo.pyx:45]: What does the `\s` do?

File: qutip/core/cy/qobjevo.pyx [hunk 4]
```diff
@@ -294,10 +282,10 @@ cdef class QobjEvo:
 
     def __call__(self, double t, dict _args=None, **kwargs):
         """
-        Get the :cls:`Qobj` at ``t``.
+        Get the :class:`~Qobj` at ``t``.
 
-        parameter
-        ---------
+        parameters
```
Comment (hodgestar) [inline qutip/core/cy/qobjevo.pyx:287]:
```suggestion
        Parameters
```
Do we normally spell parameters with a capital in these headings?

File: qutip/core/states.py [hunk 5]
```diff
@@ -1287,7 +1287,7 @@ def w_state(N=3, *, dtype=_data.Dense):
 
     Returns
     -------
-    W : qobj
+    W : :obj:`~qobj`
```
Comment (hodgestar) [inline qutip/core/states.py:1290]:
```suggestion
    W : :obj:`~Qobj`
```

File: qutip/solver/brmesolve.py [hunk 6]
```diff
@@ -52,22 +52,26 @@ def brmesolve(H, psi0, tlist, a_ops=[], e_ops=[], c_ops=[],
...
 
-    .. note:
-        ``Cubic_Spline`` have been replaced by :cls:`Coefficient`:
-            ``spline = qutip.coefficient(array, tlist=times)``
-        Whether the ``a_ops`` is time dependent is deceided by the type of the
-        operator: :cls:`Qobj` vs :cls:`QobjEvo` instead of the type of the
-        spectra.
+        .. note:
+            ``Cubic_Spline`` have been replaced by :class:`Coefficient`\:
+                ``spline = qutip.coefficient(array, tlist=times)``
```
Comment (hodgestar) [inline qutip/solver/brmesolve.py:70]:
```suggestion
            ``Cubic_Spline`` has been replaced by :class:`Coefficient`\:
                ``spline = qutip.coefficient(array, tlist=times)``
```

File: qutip/solver/brmesolve.py [hunk 7]
```diff
@@ -52,22 +52,26 @@ def brmesolve(H, psi0, tlist, a_ops=[], e_ops=[], c_ops=[],
...
+
+            Whether the ``a_ops`` is time dependent is deceided by the type of
```
Comment (hodgestar) [inline qutip/solver/brmesolve.py:72]:
```suggestion
            Whether the ``a_ops`` is time dependent is decided by the type of
```

**Review by hodgestar — CHANGES_REQUESTED** [review]
I left some small suggestions for little things I saw during the review, but thank you very much for sorting this out!

File: qutip/core/cy/qobjevo.pyx (hunk 3 above)
Comment (Ericgig) [reply qutip/core/cy/qobjevo.pyx:45]: When not there, sphinx raise a `Inline literal start-string without end-string` warning, see:
https://stackoverflow.com/questions/58633196/restructuredtext-inline-code-followed-by-s-results-in-inline-interpreted-text

**Review by hodgestar — APPROVED** [review]

File: qutip/core/cy/qobjevo.pyx (hunk 3 above)
Comment (hodgestar) [reply qutip/core/cy/qobjevo.pyx:45]: :+1:

**Comment by hodgestar** [comment]
Merged so that I can use this in the HEOM branch.
